### PR TITLE
Custom callback

### PR DESCRIPTION
**Pull request by tgallice**
This PR is relatated to #174 

All callback event receive a `TrainingState` which track metrics of the current training. And each `Callback` class are handle by the `ChainCallback`.

This implementation is light, we can add some preset params (e.g the model) in the Callback class, but I would like your thought about that ?

An other think, what do you think about an event dispatcher implementation ? There is an use case for that ? Like to add hooks in the model, trainer and other things ? 

I will add a documentation too.

### CODE REVIEW

File: tflearn/callbacks.py [hunk 1]
```diff
@@ -9,51 +9,78 @@
...
 
-    def on_epoch_end(self, **kwargs):
+    def on_batch_begin(self, training_state):
         pass
 
-    def on_sub_epoch_begin(self, **kwargs):
+    def on_sub_batch_begin(self, training_state):
         pass
 
-    def on_sub_epoch_end(self, **kwargs):
```
Comment (tgallice) [inline tflearn/callbacks.py]: @aymericdamien I have removed `on_sub_epoch_end` and `on_sub_epoch_begin` which are useless to me. It's good ?

**Comment by tgallice** [comment]
@aymericdamien Have you an example where there are many train_ops ? I didn't try this case...

**Comment by aymericdamien** [comment]
Great! That looks good :) You can test with this example from https://github.com/tflearn/tflearn/blob/master/examples/basics/logical.py#L82
```
'''
Going further: Graph combination with multiple optimizers
Create a XOR operator using product of NAND and OR operators
'''
# Data
X = [[0., 0.], [0., 1.], [1., 0.], [1., 1.]]
Y_nand = [[1.], [1.], [1.], [0.]]
Y_or = [[0.], [1.], [1.], [1.]]

# Graph definition
with tf.Graph().as_default():
    # Building a network with 2 optimizers
    g = tflearn.input_data(shape=[None, 2])
    # Nand operator definition
    g_nand = tflearn.fully_connected(g, 32, activation='linear')
    g_nand = tflearn.fully_connected(g_nand, 32, activation='linear')
    g_nand = tflearn.fully_connected(g_nand, 1, activation='sigmoid')
    g_nand = tflearn.regression(g_nand, optimizer='sgd',
                                learning_rate=2.,
                                loss='binary_crossentropy')
    # Or operator definition
    g_or = tflearn.fully_connected(g, 32, activation='linear')
    g_or = tflearn.fully_connected(g_or, 32, activation='linear')
    g_or = tflearn.fully_connected(g_or, 1, activation='sigmoid')
    g_or = tflearn.regression(g_or, optimizer='sgd',
                              learning_rate=2.,
                              loss='binary_crossentropy')
    # XOR merging Nand and Or operators
    g_xor = tflearn.merge([g_nand, g_or], mode='elemwise_mul')

    # Training
    m = tflearn.DNN(g_xor)
    m.fit(X, [Y_nand, Y_or], n_epoch=400, snapshot_epoch=False)

    # Testing
    print("Testing XOR operator")
    print("0 xor 0:", m.predict([[0., 0.]]))
    print("0 xor 1:", m.predict([[0., 1.]]))
    print("1 xor 0:", m.predict([[1., 0.]]))
print("1 xor 1:", m.predict([[1., 1.]]))
```

**Comment by aymericdamien** [comment]
It is working well, Thanks! That will be helpful.
We could also add an example or a section in 'getting_started' doc to introduce custom callbacks.

**Comment by hankwing** [comment]
Hi, 
Thanks for you great work. I just wonder would you plan to add 'model' param to **TrainingState** ? I think that is a very helpful param because sometimes people want to evaluate the model after each epoch. 

Thanks

**Pull request merged**